pkg/btcunit/rates.go

```
@@ -15,268 +15,471 @@ import (
-	// converting a fee rate to a string.
-	floatStringPrecision = 2
+	// converting a fee rate to a string. We use 3 decimal places to ensure
+	// that low fee rates (e.g., 1 sat/kvb = 0.001 sat/vbyte) are displayed
+	// with sufficient precision and not rounded to zero.
+	floatStringPrecision = 3
 )
 
-// SatPerVByte represents a fee rate in sat/vbyte. The fee rate is encoded
-// as a big.Rat to allow for fractional (sub-satoshi) fee rates.
-type SatPerVByte struct {
-	*big.Rat
-}
+var (
+	// ZeroSatPerVByte is a fee rate of 0 sat/vb.
+	ZeroSatPerVByte = NewSatPerVByte(0)
 
-// NewSatPerVByte creates a new fee rate in sat/vb. The given fee and vbytes
-// are used to calculate the fee rate.
-func NewSatPerVByte(fee btcutil.Amount, vb VByte) SatPerVByte {
-	if vb.val == 0 {
-		return SatPerVByte{big.NewRat(0, 1)}
-	}
+	// ZeroSatPerKVByte is a fee rate of 0 sat/kvb.
+	ZeroSatPerKVByte = NewSatPerKVByte(0)
+
+	// ZeroSatPerKWeight is a fee rate of 0 sat/kw.
+	ZeroSatPerKWeight = NewSatPerKWeight(0)
 
-	return SatPerVByte{
-		big.NewRat(int64(fee), safeUint64ToInt64(vb.val)),
+	// ZeroSatPerWeight is a fee rate of 0 sat/wu.
+	ZeroSatPerWeight = NewSatPerWeight(0)
+)
+
+// baseFeeRate stores the canonical representation of a fee rate, which is
+// satoshis per kilo-weight-unit (sat/kwu). All other fee rate units are
+// derived from this.
+type baseFeeRate struct {
+	// satsPerKWU is the fee rate in satoshis per kilo-weight-unit. This is
+	// the canonical representation for all fee rates within this package,
+	// chosen for its direct alignment with Bitcoin's weight unit for fee
+	// calculations and to minimize rounding errors.
+	satsPerKWU *big.Rat
+}
+
+// newBaseFeeRate creates a new baseFeeRate with the given numerator and
+// denominator. It handles the zero denominator case by returning a zero fee
+// rate.
+func newBaseFeeRate(numerator btcutil.Amount, denominator uint64) baseFeeRate {
+	if denominator == 0 {
+		return baseFeeRate{satsPerKWU: big.NewRat(0, 1)}
```

When the denominator can be zero? We can raise and propagate an error instead of swallowing it

---

This method takes an input, so it can be anything. Previously the mental model was, if the tx has 0 weight, we will pay 0 fee, so fee rate is also 0. Added a commit to panic instead.